### ISSUE TITLE
sql: require connect privilege to connect to databases

### ISSUE THREAD

**Issue by RichardJCai** [cockroachdb/cockroach#59875]
Postgres has the concept of CONNECT privilege for databases, currently cockroach does not have this.

```
CONNECT
Allows the grantee to connect to the database. This privilege is checked at connection startup (in addition to checking any restrictions imposed by pg_hba.conf).
https://www.postgresql.org/docs/13/ddl-priv.html
```

https://github.com/cockroachdb/cockroach/pull/59676 adds CONNECT privilege for databases and allows users to see metadata for objects in a db with CONNECT privilege, however we'll want to also check for CONNECT to actually be able to connect to the db

**Comment by rafiss** [comment]
cc @vy-ton @solongordon I think we should do this in 21.2. In 21.1 there will be a notice returned to the user if they `USE` a database and don't have the CONNECT privilege. To avoid breaking existing users, perhaps this change should also include something that automatically gives the CONNECT privilege if the user has a different privilege already as a one-time migration.

**Comment by solongordon** [comment]
I don't want to be too pedantic but what benefits are we looking to get out of this change? Given that CRDB allows cross-database access, it seems like "connecting" to a database doesn't have much meaning besides changing the search path. Would the main benefit be improved UX since it would prevent a user from `USE`ing a database where they don't have any privileges?

**Comment by vy-ton** [comment]
cc @piyush-singh for input too

**Comment by RichardJCai** [comment]
> I don't want to be too pedantic but what benefits are we looking to get out of this change? Given that CRDB allows cross-database access, it seems like "connecting" to a database doesn't have much meaning besides changing the search path. Would the main benefit be improved UX since it would prevent a user from `USE`ing a database where they don't have any privileges?

@solongordon I thought the plan was to get rid of cross-database references

**Comment by solongordon** [comment]
References like foreign keys, yes, but we still plan to allow cross-database access, e.g. a SELECT query which joins tables from multiple databases. See #55791 for more details.

**Comment by rafiss** [comment]
> I don't want to be too pedantic but what benefits are we looking to get out of this change? Given that CRDB allows cross-database access, it seems like "connecting" to a database doesn't have much meaning besides changing the search path. Would the main benefit be improved UX since it would prevent a user from USEing a database where they don't have any privileges?

@solongordon Yes that would be the main benefit. We've seen multiple occasions of people being confused by the UX of "being connected to a database, but not able to see anything in it." See https://github.com/cockroachdb/cockroach/issues/49313 (npgql), https://github.com/cockroachdb/cockroach/issues/37443, [this forum post about Liquibase](https://forum.cockroachlabs.com/t/user-defined-schemas-broken-with-spring-boot-and-liquibase/4136/3).

I'd also make the Postgres compatibility argument: A general rule of Postgres is that if you can CONNECT to a database, then you can view the names of the objects in it. Some of the reports we've seen above are from drivers/ORMs/tools that made this assumption with CockroachDB, which turned out not to be true.

**Comment by vy-ton** [comment]
There was a discussion [here](https://github.com/cockroachdb/cockroach/issues/56020#issuecomment-718024249) about checking `CONNECT` on cross-database queries. 

I am concerned about backwards compatibility and any auto migrations, which is why I would lean towards leaving things as is until we figure out our full story here.

**Comment by k4zuu** [comment]
I've just recently started learning CockroachDB and PostgreSQL, so if I'm wrong, please let me know.
Here is an example of a use case building a multi-tenant database where one database is given to each user

From the perspective of a multi-tenant database administrator, administrator want to reduce privileges as much as possible.
But, it is necessary for users to be able to handle at least the things related to their own database, so CREATE, DROP, CONNECT, SELECT, INSERT, DELETE, and UPDATE are assigned to user's databases.

```
CREATE DATABASE exampledb;
CREATE ROLE exampleuser WITH LOGIN PASSWORD 'exampleuser';
GRANT CREATE,DROP,CONNECT,SELECT,INSERT,DELETE,UPDATE on DATABASE exampledb TO exampleuser;
```

The psql client, when hit with a special command like `\l`, will execute it as SQL to retrieve records from pg_catalog.
With the current CockroachDB, users can see other users' databases, and they should be worried that their own information is not exposed to others.

```
$ psql ...
exampledb=> \l
                                       List of databases
    Name    |          Owner           | Encoding |  Collate   |   Ctype    | Access privileges 
------------+--------------------------+----------+------------+------------+-------------------
 defaultdb  | root                        | UTF8     | en_US.utf8 | en_US.utf8 | 
 anotherdb  | anotheruser           | UTF8     | en_US.utf8 | en_US.utf8 | 
 exampledb  | exampleuser         | UTF8     | en_US.utf8 | en_US.utf8 | 
 management | root                    | UTF8     | en_US.utf8 | en_US.utf8 | 
 postgres   | root                        | UTF8     | en_US.utf8 | en_US.utf8 | 
 system     | unknown (OID=)      | UTF8     | en_US.utf8 | en_US.utf8 | 
(6 rows)
```

For example, what if a user uses the name of a new, unreleased product as the database name?
The name of the new product will be exposed unknowingly.

What happens if the database name contains a string of sensitive information due to a mistake?
This is an extreme example, but it could lead to problems.

The above is an example of `\l`, but anyone can refer to other information from pg_catalog, which will expose information unnecessarily.

This problem is not a security problem, because things like inserting and deleting rows can be narrowed down by privileges, but it is a UX problem and can be a barrier in cases where anyone who want to serve a multi-tenant database using CockroachDB.

I do not want to understate the priority of this issue.
If an administrator thinks he or she has properly granted permissions, but unexpectedly behaves in a way that causes unnecessary questions, I believe it is just as disruptive as not checking CONNECT privilege.

I think CockroachDB is a great product and I am always helped by it.
Thank you for everything.

**Comment by tukeJonny** [comment]
I've just recently started learning CockroachDB and PostgreSQL, so if I'm wrong, please let me know.
Here is an example of a use case building a multi-tenant database where one database is given to each user

From the perspective of a multi-tenant database administrator, administrator want to reduce privileges as much as possible.
But, it is necessary for users to be able to handle at least the things related to their own database, so CREATE, DROP, CONNECT, SELECT, INSERT, DELETE, and UPDATE are assigned to user's databases.

```
CREATE DATABASE exampledb;
CREATE ROLE exampleuser WITH LOGIN PASSWORD 'exampleuser';
GRANT CREATE,DROP,CONNECT,SELECT,INSERT,DELETE,UPDATE on DATABASE exampledb TO exampleuser;
```

The psql client, when hit with a special command like `\l`, will execute it as SQL to retrieve records from pg_catalog.
With the current CockroachDB, users can see other users' databases, and they should be worried that their own information is not exposed to others.

```
$ psql ...
exampledb=> \l
                                       List of databases
    Name    |          Owner           | Encoding |  Collate   |   Ctype    | Access privileges 
------------+--------------------------+----------+------------+------------+-------------------
 defaultdb  | root                        | UTF8     | en_US.utf8 | en_US.utf8 | 
 anotherdb  | anotheruser           | UTF8     | en_US.utf8 | en_US.utf8 | 
 exampledb  | exampleuser         | UTF8     | en_US.utf8 | en_US.utf8 | 
 management | root                    | UTF8     | en_US.utf8 | en_US.utf8 | 
 postgres   | root                        | UTF8     | en_US.utf8 | en_US.utf8 | 
 system     | unknown (OID=)      | UTF8     | en_US.utf8 | en_US.utf8 | 
(6 rows)
```

For example, what if a user uses the name of a new, unreleased product as the database name?
The name of the new product will be exposed unknowingly.

What happens if the database name contains a string of sensitive information due to a mistake?
This is an extreme example, but it could lead to problems.

The above is an example of `\l`, but anyone can refer to other information from pg_catalog, which will expose information unnecessarily.

This problem is not a security problem, because things like inserting and deleting rows can be narrowed down by privileges, but it is a UX problem and can be a barrier in cases where anyone who want to serve a multi-tenant database using CockroachDB.

I do not want to understate the priority of this issue.
If an administrator thinks he or she has properly granted permissions, but unexpectedly behaves in a way that causes unnecessary questions, I believe it is just as disruptive as not checking CONNECT privilege.

I think CockroachDB is a great product and I am always helped by it.
Thank you for everything.

**Comment by rafiss** [comment]
@tukeJonny I missed your message earlier. We do currently use CONNECT privilege to determine if a user has access to view pg_catalog metadata. By the way, in Postgres, any user can see the names of all other databases. For example:

```
postgres=# create database newdb;

postgres=# revoke connect on database newdb from PUBLIC;

# login as some other non-superuser

postgres=> \l
                               List of databases
   Name    | Owner  | Encoding |   Collate   |    Ctype    | Access privileges
-----------+--------+----------+-------------+-------------+-------------------
 newdb     | rafiss | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =T/rafiss        +
           |        |          |             |             | rafiss=CTc/rafiss
 postgres  | rafiss | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 rafiss    | rafiss | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
```

Even though I cannot connect with that user:
```
$ psql 'postgresql://other_user@localhost:5432/newdb?sslmode=disable'
psql: error: connection to server at "localhost" (::1), port 5432 failed: FATAL:  permission denied for database "newdb"
DETAIL:  User does not have CONNECT privilege.
```

---

If we do implement this in the future, this is an example of the kind of migration we'd need: https://github.com/cockroachdb/cockroach/pull/72603
